plugins/redis/redissync.cpp

```diff
@@ -0,0 +1,412 @@
+    limitations under the License.
+############################################################################## */
+
+#include "platform.h"
+#include "jthread.hpp"
+#include "jhash.hpp"
+#include "eclrtl.hpp"
+#include "jstring.hpp"
+#include "redissync.hpp"
+
+namespace RedisPlugin {
+static __thread SyncConnection * cachedConnection;
+static __thread ThreadTermFunc threadHookChain;
+
+//The following class is here to ensure destruction of the cachedConnection within the main thread
+//as this is not handled by the thread hook mechanism.
+static class mainThreadCachedConnection
+{
+public :
+    mainThreadCachedConnection() { }
+    ~mainThreadCachedConnection()
+    {
+        if (cachedConnection)
+            cachedConnection->Release();
+    }
+} mainThread;
+
+static void releaseContext()
+{
+    if (cachedConnection)
+        cachedConnection->Release();
+    if (threadHookChain)
+    {
+        (*threadHookChain)();
+        threadHookChain = NULL;
+    }
+}
+
+SyncConnection::SyncConnection(ICodeContext * ctx, const char * _options, unsigned __int64 _database, const char * pswd, unsigned __int64 _timeout)
+  : Connection(ctx, _options, pswd, _timeout)
+{
+    connect(ctx, _database, pswd);
+}
+SyncConnection::SyncConnection(ICodeContext * ctx, RedisServer * _server, unsigned __int64 _database, const char * pswd)
+  : Connection(ctx, _server)
+{
+    connect(ctx, _database, pswd);
+}
+void SyncConnection::connect(ICodeContext * ctx, unsigned __int64 _database, const char * pswd)
+{
+    struct timeval to = { timeout/1000000, timeout%1000000 };
+    context = redisConnectWithTimeout(server->getIp(), server->getPort(), to);
+    assertConnection();
+    authenticate(ctx, pswd);
+    selectDB(ctx, _database);
+    init(ctx);
+}
+void SyncConnection::authenticate(ICodeContext * ctx, const char * pswd)
+{
+    if (strlen(pswd) > 0)
+    {
+        OwnedReply reply = Reply::createReply(redisCommand(context, "AUTH %b", pswd, strlen(pswd)));
+        assertOnError(reply->query(), "server authentication failed");
+    }
+}
+SyncConnection * SyncConnection::createConnection(ICodeContext * ctx, const char * options, unsigned __int64 _database, const char * pswd, unsigned __int64 _timeout)
+{
+    if (!cachedConnection)
+    {
+        cachedConnection = new SyncConnection(ctx, options, _database, pswd, _timeout);
+        threadHookChain = addThreadTermFunc(releaseContext);
+        return LINK(cachedConnection);
+    }
+
+    unsigned optionsPswdHash = hashc((const unsigned char*)options, strlen(options), hashc((const unsigned char*)pswd, strlen(pswd), 0));
+    if (cachedConnection->isSameConnection(ctx, optionsPswdHash))
+    {
+        //MORE: need to check that the connection has not expired (think hiredis REDIS_KEEPALIVE_INTERVAL is defaulted to 15s).
+        //At present updateTimeout calls assertConnection.
+        cachedConnection->updateTimeout(_timeout);
+        cachedConnection->selectDB(ctx, _database);
```

Comment:
To clarify...
One query calls a function which creates a thread for a given server and database
Another query calls a function for the same server but a different database.
If they are handled by the same actual thread (e.g., in a thread pool), then the same connection will be returned, selecting the new database.

If there was any chance of the first instance then reusing the old connection it would now access the wrong database.

I suspect in practice, because the use of the connection is limited to the get/set calls, this cannot cause problems with the sync connections.  However if it ever did persist beyond a single get/set call so the thread could switch query in between, then it would cause problems.  

I think it is a logical problem which I don't *think* can cause problems in the current usage - so wouldn't stop me merging.